ntp-proto/src/packet/mod.rs

```diff
@@ -456,7 +456,87 @@ impl<'a> NtpPacket<'a> {
                     recv_timestamp,
                     clock,
                 )),
-                efdata: Default::default(),
+                efdata: ExtensionFieldData {
+                    authenticated: vec![],
+                    encrypted: vec![],
+                    // Ignore encrypted so as not to accidentaly leak anything
+                    untrusted: input
+                        .efdata
+                        .untrusted
+                        .iter()
+                        .chain(input.efdata.authenticated.iter())
+                        .filter_map(|f| {
+                            if matches!(f, ExtensionField::UniqueIdentifier(_)) {
+                                Some(f.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .collect(),
+                },
+                mac: None,
+            },
+        }
+    }
+
+    pub fn nts_timestamp_response<C: NtpClock>(
+        system: &SystemSnapshot,
+        input: Self,
+        recv_timestamp: NtpTimestamp,
+        clock: &C,
+        cookie: &DecodedServerCookie,
+        keyset: &KeySet,
+    ) -> Self {
+        match input.header {
+            NtpHeader::V3(_) => unreachable!("NTS shouldn't work with NTPv3"),
+            NtpHeader::V4(header) => NtpPacket {
+                header: NtpHeader::V4(NtpHeaderV3V4::timestamp_response(
+                    system,
+                    header,
+                    recv_timestamp,
+                    clock,
+                )),
+                efdata: ExtensionFieldData {
+                    authenticated: input
+                        .efdata
+                        .authenticated
+                        .iter()
+                        .filter_map(|f| {
+                            if matches!(f, ExtensionField::UniqueIdentifier(_)) {
+                                Some(f.clone())
+                            } else {
+                                None
+                            }
```

Comment:
can use `into_iter` here if the `authenticated` field is moved down

Reply:
done